File: src/services/llm_service.py
"""LLM service implementation."""
from typing import List, Dict, Optional
from loguru import logger
from fastapi import HTTPException
import httpx
from datetime import datetime

class LLMService:
    """Service to handle LLM interactions."""
    
    def __init__(self, api_key: str, model: str = "deepseek-chat"):
        """Initialize LLM service."""
        self.api_key = api_key
        self.model = model
        self.api_url = "https://api.deepseek.com/v1/chat/completions"
        logger.debug(f"🤖 LLMService initialized with model {model}")
    
    async def get_chat_completion(
        self,
        messages: List[Dict[str, str]],
        temperature: float = 0.7,
        max_tokens: int = 1000
    ) -> str:
        """Get a chat completion from the LLM."""
        try:
            headers = {
                "Authorization": f"Bearer {self.api_key}",
                "Content-Type": "application/json"
            }
            
            payload = {
                "model": self.model,
                "messages": messages,
                "temperature": temperature,
                "max_tokens": max_tokens
            }
            
            logger.debug(f"🚀 Sending request to LLM with {len(messages)} messages")
            
            async with httpx.AsyncClient() as client:
                response = await client.post(
                    self.api_url,
                    headers=headers,
                    json=payload,
                    timeout=30.0
                )
                
                if response.status_code != 200:
                    raise HTTPException(
                        status_code=response.status_code,
                        detail=f"LLM API error: {response.text}"
                    )
                
                result = response.json()
                content = result["choices"][0]["message"]["content"]
                logger.debug("✅ Received response from LLM")
                
                return content
                
        except HTTPException as e:
            raise e
        except Exception as e:
            logger.error(f"❌ Error getting LLM completion: {str(e)}")
            raise HTTPException(
                status_code=500,
                detail="Failed to get response from LLM"
            )
    
    def format_chat_history(
        self,
        messages: List[Dict],
        system_prompt: Optional[str] = None
    ) -> List[Dict[str, str]]:
        """Format chat history for the LLM API."""
        formatted_messages = []
        
        # Add system prompt if provided
        if system_prompt:
            formatted_messages.append({
                "role": "system",
                "content": system_prompt
            })
        
        # Format chat messages
        for msg in messages:
            formatted_messages.append({
                "role": msg["role"],
                "content": msg["content"]
            })
        
        return formatted_messages 

File: apps/api/src/services/services/llm_service.ts
export interface ChatMessage {
    role: 'system' | 'user' | 'assistant';
    content: string;
}

export interface LLMConfig {
    temperature?: number;
    max_tokens?: number;
    model?: string;
}

export class LLMService {
    private apiKey: string;
    private model: string;
    private apiUrl: string;

    constructor(apiKey: string, model: string = "deepseek-chat") {
        this.apiKey = apiKey;
        this.model = model;
        this.apiUrl = "https://api.deepseek.com/v1/chat/completions";
        console.debug(`🤖 LLMService initialized with model ${model}`);
    }

    async getChatCompletion(
        messages: ChatMessage[],
        config: LLMConfig = {}
    ): Promise<string> {
        try {
            const headers = {
                "Authorization": `Bearer ${this.apiKey}`,
                "Content-Type": "application/json"
            };

            const payload = {
                model: config.model || this.model,
                messages,
                temperature: config.temperature ?? 0.7,
                max_tokens: config.max_tokens ?? 1000
            };

            console.debug(`🚀 Sending request to LLM with ${messages.length} messages`);

            const response = await fetch(this.apiUrl, {
                method: 'POST',
                headers,
                body: JSON.stringify(payload)
            });

            if (!response.ok) {
                throw new Error(`LLM API error: ${response.statusText}`);
            }

            const result = await response.json();
            const content = result.choices[0].message.content;
            console.debug("✅ Received response from LLM");

            return content;

        } catch (error) {
            console.error("❌ Error getting LLM completion:", error);
            throw error;
        }
    }

    formatChatHistory(
        messages: ChatMessage[],
        systemPrompt?: string
    ): ChatMessage[] {
        const formatted: ChatMessage[] = [];

        if (systemPrompt) {
            formatted.push({
                role: 'system',
                content: systemPrompt
            });
        }

        return [...formatted, ...messages];
    }
} 